substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/util/LLVMDebugLineInfo.java

```diff
@@ -0,0 +1,205 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.graal.llvm.util;
+
+import com.oracle.graal.pointsto.infrastructure.OriginalClassProvider;
+import com.oracle.graal.pointsto.infrastructure.WrappedJavaMethod;
+import com.oracle.graal.pointsto.meta.AnalysisType;
+import com.oracle.svm.core.SubstrateOptions;
+import com.oracle.svm.hosted.annotation.CustomSubstitutionMethod;
+import com.oracle.svm.hosted.annotation.CustomSubstitutionType;
+import com.oracle.svm.hosted.image.NativeImage;
+import com.oracle.svm.hosted.image.sources.SourceManager;
+import com.oracle.svm.hosted.lambda.LambdaSubstitutionType;
+import com.oracle.svm.hosted.meta.HostedMethod;
+import com.oracle.svm.hosted.meta.HostedType;
+import com.oracle.svm.hosted.substitute.InjectedFieldsType;
+import com.oracle.svm.hosted.substitute.SubstitutionMethod;
+import com.oracle.svm.hosted.substitute.SubstitutionType;
+import jdk.vm.ci.meta.LineNumberTable;
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+import jdk.vm.ci.meta.ResolvedJavaType;
+import org.graalvm.compiler.debug.DebugContext;
+import org.graalvm.compiler.graph.NodeSourcePosition;
+import org.graalvm.nativeimage.ImageSingletons;
+
+import java.nio.file.Path;
+
+import static org.graalvm.compiler.debug.GraalError.unimplemented;
+
+// Similar to NativeImageDebugLineInfo from previous releases
+public class LLVMDebugLineInfo {
+    private final int bci;
+    private final ResolvedJavaMethod method;
+    private final DebugContext debugContext;
+    private Path cachePath;
+    private Path fullFilePath;
+
+    LLVMDebugLineInfo(DebugContext debugContext, NodeSourcePosition position) {
+        int posbci = position.getBCI();
+        this.bci = (posbci >= 0 ? posbci : 0);
+        this.method = position.getMethod();
+        this.cachePath = SubstrateOptions.getDebugInfoSourceCacheRoot();
+        this.debugContext = debugContext;
+        computeFullFilePath();
+    }
+
+    private static ResolvedJavaType getOriginal(ResolvedJavaType javaType) {
+        if (javaType instanceof SubstitutionType) {
+            return ((SubstitutionType) javaType).getOriginal();
+        } else if (javaType instanceof CustomSubstitutionType<?, ?>) {
+            return ((CustomSubstitutionType<?, ?>) javaType).getOriginal();
+        } else if (javaType instanceof LambdaSubstitutionType) {
+            return ((LambdaSubstitutionType) javaType).getOriginal();
+        } else if (javaType instanceof InjectedFieldsType) {
+            return ((InjectedFieldsType) javaType).getOriginal();
+        }
+        return null;
+    }
+
+    protected static ResolvedJavaType getJavaType(HostedMethod hostedMethod, boolean wantOriginal) {
```

Comment:
I think you can get the same result by looking up `hostedMethod.getDeclaringClass().getJavaClass()`. This will be cleaner than looking into the internals of `HostedType`

Reply:
The lookup is performed this way in NativeImageDebugInfoProvider for a good reason. There are specific cases related to the use of substitutions where looking into the internals of HostedType is required in order to give a consistent view of the code base to a debugger.

The problematic situation occurs when we have a class C and substitution class S with method m1 defined by C and m2 substituted by S. We can only have one class in the debug info and we need to ensure m1 and m2 are reported as methods of that class. The current code in NativeImageDebugInfoProvider ensures that the class name is always reported as C and that the methods are reported as C.m1 and C.m2.

Reply:
Note however, that the source file for m1 needs to be C.java while the source file for m2 needs to be S.java.

Reply:
I see, that makes sense. However it would be cleaner to reuse those lookup methods from `NativeImageDebugInfoProvider` instead of copying them here.